### ISSUE TITLE
Battery health KO 

### ISSUE THREAD

**Issue by nehox** [teslamate-org/teslamate#3524]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

<img width="1777" alt="image" src="https://github.com/teslamate-org/teslamate/assets/3989362/bc9016e2-ef99-4a04-ad93-3947475a1ca9">

### Expected Behavior

in error I have this message : `Status: 500. Message: db query error: pq: invalid input syntax for type json`

SELECT  
  CASE WHEN $custom_kwh_new > 0 THEN $custom_kwh_new ELSE ('$aux'::json -> 'MaxCapacity')::text::float END as "Usable (new)", 
  ('$aux'::json -> 'CurrentCapacity')::text::float as "Usable (now)",
  ('$aux'::json -> 'CurrentCapacity')::text::float - CASE WHEN $custom_kwh_new > 0 THEN $custom_kwh_new ELSE ('$aux'::json -> 'MaxCapacity')::text::float END as "Difference"

in query $aux is empty. Anyone have the solution ?

### Steps To Reproduce

it's a fresh install, no data import. I want to try this project.

### Relevant log output

```shell
2023-11-30 13:03:59.177 UTC [48] ERROR:  invalid input syntax for type json at character 41
2023-11-30 13:03:59.177 UTC [48] DETAIL:  The input string ended unexpectedly.
2023-11-30 13:03:59.177 UTC [48] CONTEXT:  JSON data, line 1: 
2023-11-30 13:03:59.177 UTC [48] STATEMENT:  SELECT  
	  CASE WHEN 0 > 0 THEN 0 ELSE (''::json -> 'MaxRange')::text::float END as "maxrange_km",
	  (''::json -> 'CurrentRange')::text::float  as "currentrange_km",
	  CASE WHEN 0 > 0 THEN 0 ELSE (''::json -> 'MaxRange')::text::float END - (''::json -> 'CurrentRange')::text::float as "range_lost_km"
```

### Type of installation

Docker

### Version

v1.28.2

**Comment by nehox** [comment]
This request have no result :

`WITH
aux	AS
(
	SELECT cp.charge_energy_added,
		cp.car_id, (SELECT efficiency FROM cars WHERE id = $car_id) * 100.0 AS rated_efficiency,
		(cp.end_rated_range_km - cp.start_rated_range_km) AS added_range_km
	FROM charging_processes cp
		JOIN (SELECT charging_process_id, MAX(date) as date
		FROM charges
		GROUP BY charging_process_id) AS last_charges
		ON cp.id = last_charges.charging_process_id
		INNER JOIN charges c
		ON c.charging_process_id = cp.id AND c.date = last_charges.date
	WHERE cp.car_id = $car_id
		AND cp.end_date IS NOT NULL
		AND cp.end_rated_range_km > cp.start_rated_range_km
	ORDER BY cp.end_date DESC 
	LIMIT 1
),  
CurrentCapacity	 AS
(
	SELECT AVG(Capacity) AS CurrentCapacity FROM
  (SELECT (100.0 * cp.charge_energy_added) / (GREATEST(1,MAX(usable_battery_level) - MIN(usable_battery_level))) AS Capacity	
    FROM charging_processes cp
	  INNER JOIN charges c	ON cp.id = c.charging_process_id
    INNER JOIN aux ON cp.car_id = aux.car_id
	  WHERE cp.car_id = $car_id AND cp.charge_energy_added >= aux.rated_efficiency 
    GROUP BY  cp.charge_energy_added, cp.end_date
    ORDER BY cp.end_date DESC 
    LIMIT 5
  ) AS lastEstimatedCapacity
), 
MaxCapacity AS
(
   SELECT AVG(c.rated_battery_range_km * aux.rated_efficiency  / c.usable_battery_level) AS MaxCapacity, cp.id
   FROM charging_processes cp
	JOIN (SELECT charging_process_id, MAX(date) as date FROM charges 
                    WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges	
        ON cp.id = last_charges.charging_process_id
	INNER JOIN charges c ON c.charging_process_id = cp.id AND c.date = last_charges.date
	INNER JOIN positions p ON p.id = cp.position_id
        INNER JOIN aux ON cp.car_id = aux.car_id
	WHERE cp.car_id = $car_id
		AND cp.end_date IS NOT NULL
		AND cp.charge_energy_added >= aux.rated_efficiency 
	GROUP BY cp.id 
        ORDER BY 1 DESC 
        LIMIT 1
), 
CurrentRange AS
(
    SELECT
		floor(extract(epoch from date)/86400)*86400 AS timecurrent,
		sum(rated_battery_range_km) / sum(usable_battery_level) * 100 AS rated_range_km
	FROM (
		SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c 
		JOIN charging_processes p ON p.id = c.charging_process_id 
		WHERE p.car_id = $car_id AND usable_battery_level IS NOT NULL) AS data
	GROUP BY 1
	ORDER BY 1 DESC
	LIMIT 1
), 
MaxRange AS
(
    SELECT
		floor(extract(epoch from date)/86400)*86400 AS time,
		sum(rated_battery_range_km) / sum(usable_battery_level) * 100 AS max_rated_range_km
	FROM (
		SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c 
		JOIN charging_processes p ON p.id = c.charging_process_id 
		WHERE p.car_id = $car_id AND usable_battery_level IS NOT NULL) AS data
	GROUP BY 1
	ORDER BY 2 DESC
	LIMIT 1
) 
SELECT  CONCAT('{"LastChargekWhAdded": ', aux.charge_energy_added, 
  ', "LastMileageAdded" : ', convert_km(aux.added_range_km,'$length_unit'),
  ', "MaxRange": ', convert_km(MaxRange.max_rated_range_km,'$length_unit'),
  ', "CurrentRange": ',convert_km(CurrentRange.rated_range_km,'$length_unit'),
  ', "MaxCapacity": ', MaxCapacity.MaxCapacity,
  ', "CurrentCapacity": ', CASE WHEN CurrentCapacity.CurrentCapacity IS NULL THEN 1 ELSE CurrentCapacity.CurrentCapacity END,
  ', "RatedEfficiency": ',aux.rated_efficiency, '}') 
FROM MaxRange, CurrentRange, Aux, MaxCapacity, CurrentCapacity; `

**Comment by nehox** [comment]
After a long charge, 33% to 80%. Dashboard is ok.

**Comment by JakobLichterfeld** [comment]
Thanks for reporting. Battery Health do need data logged before it can calculate the battery health.
You can see hints about it while hovering, for example:
![image](https://github.com/teslamate-org/teslamate/assets/16510328/5300685b-9bd0-44c0-ae0f-402178057602)

But anyway, the behavior when no data is available can be improved. Perhaps @swiffer or @oivindoh can look into it.

**Comment by cwanja** [comment]
Also comment that this might use the consumption value and it is documented that you need a certain amount of charge cycles to derive that value. 

https://docs.teslamate.org/docs/faq#why-are-no-consumption-values-displayed-in-grafana

@jheredianet can confirm further, I believe this is merged from your dashboards.

**Comment by ZevaKikel** [comment]
> Thanks for reporting. Battery Health do need data logged before it can calculate the battery health. You can see hints about it while hovering, for example: ![image](https://private-user-images.githubusercontent.com/16510328/288057048-5300685b-9bd0-44c0-ae0f-402178057602.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMDgzMDksIm5iZiI6MTcwMjMwODAwOSwicGF0aCI6Ii8xNjUxMDMyOC8yODgwNTcwNDgtNTMwMDY4NWItOWJkMC00NGMwLWFlMGYtNDAyMTc4MDU3NjAyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjExVDE1MjAwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3ZGNmMGNhZDkxYzQ3NDgwZmU1ZjA1NGJmM2EyZTQxODI2NjQxYTczNDBlMGQ2OGUwMjE1NDFiZTk1MmM3ODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EnijL4TlJ10UB3pjWQBGcJ8qwFRsDT2IRVS31MEhl7o)
> 
> But anyway, the behavior when no data is available can be improved. Perhaps @swiffer or @oivindoh can look into it.

I have the same problem, but in my case I have 5 charging sessions already logged with Teslamate:
![image](https://github.com/teslamate-org/teslamate/assets/102579946/1c871565-3c4e-483b-8b13-8ba16ea2ec5f)

The Battery health (both the standard and the LFP) shows "No Data". Other dashboards works fine
![image](https://github.com/teslamate-org/teslamate/assets/102579946/26906c15-e5cf-42b4-b708-b7b399542780)

**Comment by swiffer** [comment]
LFP variants of dashboards have been added. However for battery health efficieny is only calculated if charged for min 10 minutes and below 95% currently. this needs to be changed in teslamate (elixir) codebase

https://github.com/teslamate-org/teslamate/blob/83a0b38854572ba8fc816cd6df5bac9b71c00a01/lib/teslamate/log.ex#L603

i question if we can simply raise the upper bound or if we need to do that dynamically based on the yet to be introduced vehicle type battery config setting proposed by @NirKli here

https://github.com/teslamate-org/teslamate/issues/3438#issuecomment-1835618834

@NirKli - it shouldn't be a overall setting but configured per vehicle, right? we might need to implement a flow similar to how charge costs are set to be able to set car specific settings? (add a battery type single stat chart to battery health that when clicking on it opens a settings page for the currently selected car in teslamate that then updates a value in the cars table - changing it should trigger recalculate_efficiency afterwards). 

@JakobLichterfeld - what do you think?

**Comment by JakobLichterfeld** [comment]
> i question if we can simply raise the upper bound or if we need to do that dynamically based on the yet to be introduced vehicle type battery config setting proposed by @NirKli here

From my point of view we need to do that dynamically, as battery health calculation for non LFP variants will break otherwise, as there values reported for 100% charging are not accurate for those battery types.

**Comment by danggeni** [comment]
I checked the data in the same situation, 

PostgreSQL did not have the efficiency information for the cars table. 

I looked at the source code of teslamate and the problem was that it was updating the efficiency information based on the information up to 95% charge.
I also have a ModelY RWD, so I charge it to 100% every charging session, so the data was not updated, and the battery information in Grafana was always 'No Data'. 

I checked the source code and charged the car from 70% charge to about 10% and got the results.

It would be nice to allow vehicles with LFP batteries to reach 100% to calculate results.

![Uploading IMG_2005.jpg
![IMG_2004](https://github.com/teslamate-org/teslamate/assets/10123843/1116c76d-44f5-4a96-a206-1598352feb3e)
![IMG_2003](https://github.com/teslamate-org/teslamate/assets/10123843/fdc8b3bd-3e0e-4dde-8179-6092c8f25f1b)
…]()

**Comment by inigoml** [comment]
Fix here:
https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/pull/63

**Comment by inigoml** [comment]
PR created. If approved, will be fixed for next release.

**Comment by Viald** [comment]
With the **1.28.2** release, I have new LFP dashboards but the issue is still existing.
![Image 1142](https://github.com/teslamate-org/teslamate/assets/3048030/f0c7458d-84d0-40c5-9ff5-95cbfb056a7e)
In which release this patch should be included ?

**Comment by inigoml** [comment]
This patch was integrated into dashboard author repository and lately created a new PR for Teslamate. But not yet approved.
Anyway, we fixed Battery Dashboard but not LFP dashboard. But the problem is probably the same.

You can open source file and just copy & paste code into Grafana dashboard and will be available immediately or wait until it's included in Teslamate master and a new release is launched.

**Comment by Viald** [comment]
I have updated Teslamate-CustomGrafanaDashboards from repository. Now BatteryHealth.json release is the last one.
But I still have the same issue

**Comment by jheredianet** [comment]
From the error you show, it looks like it's Grafana that isn't being able to access the database (400 Bad Request).
Are you sure you have imported the dashboard with its corresponding datasource?
Yesterday was approved this PR https://github.com/teslamate-org/teslamate/pull/3593
Maybe you should wait to the release that it would be integrated as all the common dashboards.

**Comment by Viald** [comment]
Yes I'm sure. That's weird, I have this error message from Chrome console for each grafana dashboard even if there is no visual error.

**Comment by JakobLichterfeld** [comment]
Did you have any data captured with TeslaMate yet?

**Comment by Viald** [comment]
Yes
![image](https://github.com/teslamate-org/teslamate/assets/3048030/2739fcf2-e716-428d-b388-cdba89b24d23)

**Comment by jheredianet** [comment]
@Viald  may I kindly ask you to import this release in order make some tests?  https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/blob/develop/dashboards/BatteryHealth.json

I've made some adjustments to control none o null values (just on the Developer branch for now). Once you import this Dasboard, you need to go to "Dashboard settings" | Variables the click on "aux" variable and let me know the values you get in the Preview section, as the screenshot below:
![image](https://github.com/teslamate-org/teslamate/assets/45735191/00065b25-2a3c-46df-a275-996cbd1097a2)

**Comment by Viald** [comment]
I did it, and found None in the Preview section.
![image](https://github.com/teslamate-org/teslamate/assets/3048030/180d3bbe-ef31-4734-8edd-103dd076368b)

**Comment by jheredianet** [comment]
Ok, it's because there is some query that it is not working based on the data you have. 
May you run these queries individually? 
Let me know all the results (a screenshot would be ok) in order to see exactly which one is not getting the expected data.
I'm assuming that your car id is 1, if not, change it accordantly in each query.


To get DerivatedEfficiency:
``
SELECT (charge_energy_added / NULLIF(end_rated_range_km - start_rated_range_km, 0))::numeric * 100.0 as "efficiency",
        count(*) as count FROM  charging_processes 
    WHERE car_id = 1 
        AND duration_min > 10  
        AND end_battery_level <= 95  
        AND start_rated_range_km IS NOT NULL  
        AND end_rated_range_km IS NOT NULL  
        AND charge_energy_added > 0
    GROUP BY  1
    ORDER BY  2 DESC
    LIMIT 1
``

Last info on charges:
``
SELECT cp.charge_energy_added,
		cp.car_id, 
		(cp.end_rated_range_km - cp.start_rated_range_km) AS added_range_km
	FROM charging_processes cp
		JOIN (SELECT charging_process_id, MAX(date) as date
		FROM charges
		GROUP BY charging_process_id) AS last_charges
		ON cp.id = last_charges.charging_process_id
		INNER JOIN charges c
		ON c.charging_process_id = cp.id AND c.date = last_charges.date
	WHERE cp.car_id = 1
		AND cp.end_date IS NOT NULL
		AND cp.end_rated_range_km > cp.start_rated_range_km
	ORDER BY cp.end_date DESC 
	LIMIT 1
``

Current Capacity:
``
SELECT AVG(Capacity) AS CurrentCapacity FROM
    (SELECT (100.0 * cp.charge_energy_added) / (GREATEST(1,MAX(usable_battery_level) - MIN(usable_battery_level))) AS Capacity	
    FROM charging_processes cp
	  INNER JOIN charges c ON cp.id = c.charging_process_id
    INNER JOIN aux ON cp.car_id = 1
	  WHERE cp.car_id = 1 AND cp.charge_energy_added >= 10 
    GROUP BY  cp.charge_energy_added, cp.end_date
    ORDER BY cp.end_date DESC 
    LIMIT 5
  ) AS lastEstimatedCapacity
``

Max Capacity:
``
   SELECT AVG(c.rated_battery_range_km * 14  / c.usable_battery_level) AS MaxCapacity, cp.id
   FROM charging_processes cp
	JOIN (SELECT charging_process_id, MAX(date) as date FROM charges 
                    WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges	
        ON cp.id = last_charges.charging_process_id
	INNER JOIN charges c ON c.charging_process_id = cp.id AND c.date = last_charges.date
	INNER JOIN positions p ON p.id = cp.position_id
        INNER JOIN aux ON cp.car_id = 1
	WHERE cp.car_id = 1
		AND cp.end_date IS NOT NULL
		AND cp.charge_energy_added >= 14
	GROUP BY cp.id 
        ORDER BY 1 DESC 
        LIMIT 1
``

Current Range:
``
SELECT
		floor(extract(epoch from date)/86400)*86400 AS timecurrent,
		CASE WHEN sum(usable_battery_level) = 0 THEN sum(rated_battery_range_km) * 100
		     ELSE sum(rated_battery_range_km) / sum(usable_battery_level) * 100
        END AS rated_range_km
	FROM (
		SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c 
		JOIN charging_processes p ON p.id = c.charging_process_id 
		WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data
	GROUP BY 1
	ORDER BY 1 DESC
	LIMIT 1
``

Max Range:
``
    SELECT
		floor(extract(epoch from date)/86400)*86400 AS time,
	    CASE WHEN sum(usable_battery_level) = 0 THEN sum(rated_battery_range_km) * 100
		     ELSE sum(rated_battery_range_km) / sum(usable_battery_level) * 100
	    END AS max_rated_range_km
    FROM (
		SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c 
		JOIN charging_processes p ON p.id = c.charging_process_id 
		WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data
	GROUP BY 1
	ORDER BY 2 DESC
	LIMIT 1
``

**Comment by Viald** [comment]
Sorry, I'm not a grafana expert, I did my best. Is this result below is readable ?

I saw 2 errors as well:
![image](https://github.com/teslamate-org/teslamate/assets/3048030/1d884418-762d-4adb-bfda-33f96adf0dce)


{
  "request": {
    "url": "api/ds/query?ds_type=postgres&requestId=Q117",
    "method": "POST",
    "data": {
      "queries": [
        {
          "refId": "DerivatedEfficiency",
          "datasource": {
            "type": "postgres",
            "uid": "PC98BA2F4D77E1A42"
          },
          "rawSql": "SELECT (charge_energy_added / NULLIF(end_rated_range_km - start_rated_range_km, 0))::numeric * 100.0 as \"efficiency\", count(*) as count FROM  charging_processes  WHERE car_id = 1  AND duration_min > 10   AND end_battery_level <= 95   AND start_rated_range_km IS NOT NULL   AND end_rated_range_km IS NOT NULL   AND charge_energy_added > 0 GROUP BY  1 ORDER BY  2 DESC LIMIT 1",
          "format": "table",
          "datasourceId": 1,
          "intervalMs": 60000,
          "maxDataPoints": 1180
        },
        {
          "refId": "Last info on charges",
          "datasource": {
            "type": "postgres",
            "uid": "PC98BA2F4D77E1A42"
          },
          "rawSql": "SELECT cp.charge_energy_added, cp.car_id,  (cp.end_rated_range_km - cp.start_rated_range_km) AS added_range_km FROM charging_processes cp JOIN (SELECT charging_process_id, MAX(date) as date FROM charges GROUP BY charging_process_id) AS last_charges ON cp.id = last_charges.charging_process_id INNER JOIN charges c ON c.charging_process_id = cp.id AND c.date = last_charges.date WHERE cp.car_id = 1 AND cp.end_date IS NOT NULL AND cp.end_rated_range_km > cp.start_rated_range_km ORDER BY cp.end_date DESC  LIMIT 1",
          "format": "table",
          "datasourceId": 1,
          "intervalMs": 60000,
          "maxDataPoints": 1180
        },
        {
          "refId": "Current Capacity",
          "datasource": {
            "type": "postgres",
            "uid": "PC98BA2F4D77E1A42"
          },
          "rawSql": "SELECT AVG(Capacity) AS CurrentCapacity FROM (SELECT (100.0 * cp.charge_energy_added) / (GREATEST(1,MAX(usable_battery_level) - MIN(usable_battery_level))) AS Capacity\t FROM charging_processes cp INNER JOIN charges c ON cp.id = c.charging_process_id INNER JOIN aux ON cp.car_id = 1 WHERE cp.car_id = 1 AND cp.charge_energy_added >= 10  GROUP BY  cp.charge_energy_added, cp.end_date ORDER BY cp.end_date DESC  LIMIT 5 ) AS lastEstimatedCapacity",
          "format": "table",
          "datasourceId": 1,
          "intervalMs": 60000,
          "maxDataPoints": 1180
        },
        {
          "refId": "Max Capacity",
          "datasource": {
            "type": "postgres",
            "uid": "PC98BA2F4D77E1A42"
          },
          "rawSql": "SELECT AVG(c.rated_battery_range_km * 14  / c.usable_battery_level) AS MaxCapacity, cp.id FROM charging_processes cp JOIN (SELECT charging_process_id, MAX(date) as date FROM charges  WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges\t ON cp.id = last_charges.charging_process_id INNER JOIN charges c ON c.charging_process_id = cp.id AND c.date = last_charges.date INNER JOIN positions p ON p.id = cp.position_id INNER JOIN aux ON cp.car_id = 1 WHERE cp.car_id = 1 AND cp.end_date IS NOT NULL AND cp.charge_energy_added >= 14 GROUP BY cp.id  ORDER BY 1 DESC  LIMIT 1",
          "format": "table",
          "datasourceId": 1,
          "intervalMs": 60000,
          "maxDataPoints": 1180
        },
        {
          "refId": "Current Range",
          "datasource": {
            "type": "postgres",
            "uid": "PC98BA2F4D77E1A42"
          },
          "rawSql": "SELECT floor(extract(epoch from date)/86400)*86400 AS timecurrent, CASE WHEN sum(usable_battery_level) = 0 THEN sum(rated_battery_range_km) * 100 ELSE sum(rated_battery_range_km) / sum(usable_battery_level) * 100 END AS rated_range_km FROM ( SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c  JOIN charging_processes p ON p.id = c.charging_process_id  WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data GROUP BY 1 ORDER BY 1 DESC LIMIT 1",
          "format": "table",
          "datasourceId": 1,
          "intervalMs": 60000,
          "maxDataPoints": 1180
        },
        {
          "refId": "Max Range",
          "datasource": {
            "type": "postgres",
            "uid": "PC98BA2F4D77E1A42"
          },
          "rawSql": "SELECT floor(extract(epoch from date)/86400)*86400 AS time, CASE WHEN sum(usable_battery_level) = 0 THEN sum(rated_battery_range_km) * 100 ELSE sum(rated_battery_range_km) / sum(usable_battery_level) * 100 END AS max_rated_range_km FROM ( SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c  JOIN charging_processes p ON p.id = c.charging_process_id  WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data GROUP BY 1 ORDER BY 2 DESC LIMIT 1",
          "format": "table",
          "datasourceId": 1,
          "intervalMs": 60000,
          "maxDataPoints": 1180
        }
      ],
      "from": "1705818740662",
      "to": "1705840340662"
    },
    "hideFromInspector": false
  },
  "response": {
    "results": {
      "Current Capacity": {
        "error": "db query error: pq: relation \"aux\" does not exist",
        "status": 500,
        "frames": [
          {
            "schema": {
              "refId": "Current Capacity",
              "meta": {
                "typeVersion": [
                  0,
                  0
                ],
                "executedQueryString": "SELECT AVG(Capacity) AS CurrentCapacity FROM (SELECT (100.0 * cp.charge_energy_added) / (GREATEST(1,MAX(usable_battery_level) - MIN(usable_battery_level))) AS Capacity\t FROM charging_processes cp INNER JOIN charges c ON cp.id = c.charging_process_id INNER JOIN aux ON cp.car_id = 1 WHERE cp.car_id = 1 AND cp.charge_energy_added >= 10  GROUP BY  cp.charge_energy_added, cp.end_date ORDER BY cp.end_date DESC  LIMIT 5 ) AS lastEstimatedCapacity"
              },
              "fields": []
            },
            "data": {
              "values": []
            }
          }
        ],
        "refId": "Current Capacity"
      },
      "Current Range": {
        "status": 200,
        "frames": [
          {
            "schema": {
              "refId": "Current Range",
              "meta": {
                "typeVersion": [
                  0,
                  0
                ],
                "executedQueryString": "SELECT floor(extract(epoch from date)/86400)*86400 AS timecurrent, CASE WHEN sum(usable_battery_level) = 0 THEN sum(rated_battery_range_km) * 100 ELSE sum(rated_battery_range_km) / sum(usable_battery_level) * 100 END AS rated_range_km FROM ( SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c  JOIN charging_processes p ON p.id = c.charging_process_id  WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data GROUP BY 1 ORDER BY 1 DESC LIMIT 1"
              },
              "fields": [
                {
                  "name": "timecurrent",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64",
                    "nullable": true
                  }
                },
                {
                  "name": "rated_range_km",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64",
                    "nullable": true
                  }
                }
              ]
            },
            "data": {
              "values": [
                [
                  1705449600
                ],
                [
                  400.43157134098277
                ]
              ]
            }
          }
        ],
        "refId": "Current Range"
      },
      "DerivatedEfficiency": {
        "status": 200,
        "frames": [
          {
            "schema": {
              "refId": "DerivatedEfficiency",
              "meta": {
                "typeVersion": [
                  0,
                  0
                ],
                "executedQueryString": "SELECT (charge_energy_added / NULLIF(end_rated_range_km - start_rated_range_km, 0))::numeric * 100.0 as \"efficiency\", count(*) as count FROM  charging_processes  WHERE car_id = 1  AND duration_min > 10   AND end_battery_level <= 95   AND start_rated_range_km IS NOT NULL   AND end_rated_range_km IS NOT NULL   AND charge_energy_added > 0 GROUP BY  1 ORDER BY  2 DESC LIMIT 1"
              },
              "fields": []
            },
            "data": {
              "values": []
            }
          }
        ],
        "refId": "DerivatedEfficiency"
      },
      "Last info on charges": {
        "status": 200,
        "frames": [
          {
            "schema": {
              "refId": "Last info on charges",
              "meta": {
                "typeVersion": [
                  0,
                  0
                ],
                "executedQueryString": "SELECT cp.charge_energy_added, cp.car_id,  (cp.end_rated_range_km - cp.start_rated_range_km) AS added_range_km FROM charging_processes cp JOIN (SELECT charging_process_id, MAX(date) as date FROM charges GROUP BY charging_process_id) AS last_charges ON cp.id = last_charges.charging_process_id INNER JOIN charges c ON c.charging_process_id = cp.id AND c.date = last_charges.date WHERE cp.car_id = 1 AND cp.end_date IS NOT NULL AND cp.end_rated_range_km > cp.start_rated_range_km ORDER BY cp.end_date DESC  LIMIT 1"
              },
              "fields": [
                {
                  "name": "charge_energy_added",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64",
                    "nullable": true
                  }
                },
                {
                  "name": "car_id",
                  "type": "number",
                  "typeInfo": {
                    "frame": "int16",
                    "nullable": true
                  }
                },
                {
                  "name": "added_range_km",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64",
                    "nullable": true
                  }
                }
              ]
            },
            "data": {
              "values": [
                [
                  47.64
                ],
                [
                  1
                ],
                [
                  373.99
                ]
              ]
            }
          }
        ],
        "refId": "Last info on charges"
      },
      "Max Capacity": {
        "error": "db query error: pq: relation \"aux\" does not exist",
        "status": 500,
        "frames": [
          {
            "schema": {
              "refId": "Max Capacity",
              "meta": {
                "typeVersion": [
                  0,
                  0
                ],
                "executedQueryString": "SELECT AVG(c.rated_battery_range_km * 14  / c.usable_battery_level) AS MaxCapacity, cp.id FROM charging_processes cp JOIN (SELECT charging_process_id, MAX(date) as date FROM charges  WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges\t ON cp.id = last_charges.charging_process_id INNER JOIN charges c ON c.charging_process_id = cp.id AND c.date = last_charges.date INNER JOIN positions p ON p.id = cp.position_id INNER JOIN aux ON cp.car_id = 1 WHERE cp.car_id = 1 AND cp.end_date IS NOT NULL AND cp.charge_energy_added >= 14 GROUP BY cp.id  ORDER BY 1 DESC  LIMIT 1"
              },
              "fields": []
            },
            "data": {
              "values": []
            }
          }
        ],
        "refId": "Max Capacity"
      },
      "Max Range": {
        "status": 200,
        "frames": [
          {
            "schema": {
              "refId": "Max Range",
              "meta": {
                "typeVersion": [
                  0,
                  0
                ],
                "executedQueryString": "SELECT floor(extract(epoch from date)/86400)*86400 AS time, CASE WHEN sum(usable_battery_level) = 0 THEN sum(rated_battery_range_km) * 100 ELSE sum(rated_battery_range_km) / sum(usable_battery_level) * 100 END AS max_rated_range_km FROM ( SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c  JOIN charging_processes p ON p.id = c.charging_process_id  WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data GROUP BY 1 ORDER BY 2 DESC LIMIT 1"
              },
              "fields": [
                {
                  "name": "time",
                  "type": "time",
                  "typeInfo": {
                    "frame": "time.Time",
                    "nullable": true
                  }
                },
                {
                  "name": "max_rated_range_km",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64",
                    "nullable": true
                  }
                }
              ]
            },
            "data": {
              "values": [
                [
                  1702944000000
                ],
                [
                  401.64309704790526
                ]
              ]
            }
          }
        ],
        "refId": "Max Range"
      }
    }
  }
}

**Comment by jheredianet** [comment]
Hi @Viald It's hard to read those long lines in JSON. 
In the other hand, I've edited the queries in the previous post (the ones that failed). 


If you are using Grafana, you may go to main menu "Explore", then choose your data source, then click on "code", paste and run the query as the image below (the results will be showed at the button):

![image](https://github.com/teslamate-org/teslamate/assets/45735191/6324a18c-07c9-421b-ad00-7bfd365b60cb)

Could you run them again? and show a screenshot for each one?

**Comment by Viald** [comment]
Sure, here it is.

![image](https://github.com/teslamate-org/teslamate/assets/3048030/5f622b5f-c4de-4a01-9e30-4e4e1f378ef8)
![image](https://github.com/teslamate-org/teslamate/assets/3048030/97fba459-149e-4a55-8d40-81e116af7b6d)
![image](https://github.com/teslamate-org/teslamate/assets/3048030/ee4dcacd-41f2-4b1f-ae86-1300e6e017ff)
![image](https://github.com/teslamate-org/teslamate/assets/3048030/c8c2a578-3395-4a55-a87c-f4c4fb7565df)
![image](https://github.com/teslamate-org/teslamate/assets/3048030/1649d425-eee0-45ed-96c5-9c23684c22b7)
![image](https://github.com/teslamate-org/teslamate/assets/3048030/689563a9-6ee1-4dd0-ba09-7ed6bac85fe6)

**Comment by jheredianet** [comment]
Ok @Viald, so the issue is something with your data. We cannot get the current efficiency with this query:

``
SELECT (charge_energy_added / NULLIF(end_rated_range_km - start_rated_range_km, 0))::numeric * 100.0 as "efficiency", count(*) as count FROM  charging_processes  WHERE car_id = 1  AND duration_min > 10   AND end_battery_level <= 95   AND start_rated_range_km IS NOT NULL   AND end_rated_range_km IS NOT NULL   AND charge_energy_added > 0 GROUP BY  1 ORDER BY  2 DESC LIMIT 1`
``

This is strange, it looks like you don't have any charge that lasts longer than 10 minutes and is less than 95%. 
As the graph you shared before, it says that you have logged 2.181 Kms in Teslamate and you have added 745kWh and only with AC, not DC o SuperCharger, and (is that true)? Which model is your car? Have you imported data from other sources?

Maybe you should check your charging_processes  table and check if you have unclosed sessions (here you can see how to manually fix if it is your case https://docs.teslamate.org/docs/maintenance/manually_fixing_data). 


Anyway, could you go to the Efficiency Dashboard and see what do you get on "Current Rated Efficiency" as the image below?

![image](https://github.com/teslamate-org/teslamate/assets/45735191/ed613e40-67eb-4f17-98e2-d5f506d2733b)

If you want, you may share a screenshot of the "Charges Dashboard" as well to clarify what is happening.

**Comment by Viald** [comment]
> This is strange, it looks like you don't have any charge that lasts longer than 10 minutes.

All charges take few hours (see below)

> and is less than 95%. 

True, as recommanded by Tesla for FLP battery, I always charge at 100%

>As the graph you shared before, it says that you have logged 2.181 Kms in Teslamate and you have added 745kWh and only with AC, not DC o SuperCharger, is that true?

Since I have been using Teslamate, I have not charged the car with supercharger, only AC.

> Which model is your car?

Model 3 2021 (LFP battery)

>Have you imported data from other sources?

No

> Anyway, could you share what you get on the "Efficiency Dashboard" as the image below?

![image](https://github.com/teslamate-org/teslamate/assets/3048030/2ddb254a-986f-4545-9eab-afccb8978174)

> If you want, you may share a screenshot of the "Charges Dashboard" as well to clarify what is happening.

![image](https://github.com/teslamate-org/teslamate/assets/3048030/a7c5d80d-7e79-4488-99c4-e853ccbe6e19)

**Comment by JakobLichterfeld** [comment]
So why are you not using LFP dashboard?

**Comment by Viald** [comment]
As you specify, I have the same issue with the LFP dashboard.
![image](https://github.com/teslamate-org/teslamate/assets/3048030/69eea415-8f3a-4bf4-99e6-2f6985057bfc)

**Comment by jheredianet** [comment]
As I've told you before, we cannot get the efficiency, as you can see in the screen shot. Any efficiency at all

> ![image](https://private-user-images.githubusercontent.com/3048030/298532322-2ddb254a-986f-4545-9eab-afccb8978174.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5MTc1MDUsIm5iZiI6MTcwNTkxNzIwNSwicGF0aCI6Ii8zMDQ4MDMwLzI5ODUzMjMyMi0yZGRiMjU0YS05ODZmLTQ1NDUtOWVhYi1hZmNjYjg5NzgxNzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjJUMDk1MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VmNDVhMzhjYjJmMzZjMmM5YmIzOTUxYzIwYTVmMThkNjk1MmJjNjM5NGIzYmVjOWFlMDg0MjRjNDYwMGJmMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.BKC7u4KzjaAVClSosumNmnzurqpFVypLj5GDX8qeInI)

I recommend you to charge your car at different percent, once to 50% other to 60% .... 90%, etc. Not always at 100% except you strictly needed. I don't think changing always to 100% is a good way to keep the battery health even for a LFP one. As you show in your graph your charges are always to 100%

> ![image](https://private-user-images.githubusercontent.com/3048030/298532932-a7c5d80d-7e79-4488-99c4-e853ccbe6e19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5MTc1MDUsIm5iZiI6MTcwNTkxNzIwNSwicGF0aCI6Ii8zMDQ4MDMwLzI5ODUzMjkzMi1hN2M1ZDgwZC03ZTc5LTQ0ODgtOTljNC1lODUzY2NiZTZlMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjJUMDk1MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzM3MzNiYTdkOWFhNDZiYzc1MDExNjNjMDYyZWIwNzc0M2U5Y2E4NDBlZTJiZjk5YzAwZGI1Zjk0YTQyMjhkZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.WcCfEBIteuJVlPM3LP0uwRHGh9pO3B25URpLvJKJgMs)

**Comment by Viald** [comment]
Ok thank you. I'll charge like you suggest and let you know if something change in Teslamate. Regarding the best way to charge LFP battery, I read a lot, not sure that someone has the truth. So I ask Tesla directly who recommends to charge at 100% as often as possible, but yes, I should do it many times a week.

**Comment by JakobLichterfeld** [comment]
> I'll charge like you suggest and let you know if something change in Teslamate

Please stick to what your Tesla is suggesting for your vehicle ind charge view in vehicle and do not follow others. Your battery health must be really fine, as you do not use DC charging and do have many times a charger connected when not loading, so the battery sensors can recalculate.
I do not have a LFP vehicle.

**Comment by jheredianet** [comment]
> > I don't think changing always to 100% is a good way to keep the battery health even for a LFP one.
> 
> Please only post valid informarion.

Do you really think this information is invalid? It's only an opinion, as valid as yours.

As far as I know there is no absolute truth in this (as @Viald commented). Even if you call and ask Tesla it will depend on the person who assists you and will give you different information.
Also, if that were the case, why doesn't Tesla put it in writing in the user manual?

I don't have an LFP battery either, but I know many users at the Tesla Club in my city with this type of battery with different use cases and they all agree that they charge at least once a week to 100%, **_but not always_**.

As I said, this is just an opinion and there is no absolute truth in this. 
What we all agree on is that LFP batteries are less sensitive to degradation compared to other batteries with different chemistry.

**Comment by Viald** [comment]
On Tesla app, Tesla provides information about the way to charge your car. When you reduce the charge rate under 100%, a message appears telling that you should keep the limite at 100% and do a complete charge at least once a week.
We can tell that it's official.

**Comment by Viald** [comment]
> TeslaMate was developed before LFP batteries were included, therefore we do have a limit for the calculation, as 2170 NCA cells behave very different and calculation at 100% charge is 100% off. As wrote above, we need to change this for LFP variants, and are searching ways to determine the battery type automatically. Only if there is a robust solution, we will update it.

For my knowledge, battery health is calculated by Teslamate and extracted by Grafana, or Teslamate only collects data from Tesla servers and Grafana computes data to calculate battery health ?
Thanks

**Comment by JakobLichterfeld** [comment]
Battery health is not available via API and therefore not logged. In this case Grafana Dashboard use all the available data from TeslaMate to do an estimate calculation of the battery health based on charges, rated capacity, actual capacity etc.

**Comment by Viald** [comment]
Ok, so pending the solution to determine the battery type automatically, what about having dedicated dashboards for the LFP battery which take into account the fact that we (LFP owners) often charge at 100% ?

**Comment by JakobLichterfeld** [comment]
That's where the efficiency calculation comes into place, which is done in TeslaMate and only if you charge below 95% as stated out earlier: https://github.com/teslamate-org/teslamate/issues/3524#issuecomment-1850452981

**Comment by swiffer** [comment]
> LFP variants of dashboards have been added. However for battery health efficieny is only calculated if charged for min 10 minutes and below 95% currently. this needs to be changed in teslamate (elixir) codebase
> 
> https://github.com/teslamate-org/teslamate/blob/83a0b38854572ba8fc816cd6df5bac9b71c00a01/lib/teslamate/log.ex#L603
> 
> i question if we can simply raise the upper bound or if we need to do that dynamically based on the yet to be introduced vehicle type battery config setting proposed by @NirKli here
> 
> https://github.com/teslamate-org/teslamate/issues/3438#issuecomment-1835618834
> 
> @NirKli - it shouldn't be a overall setting but configured per vehicle, right? we might need to implement a flow similar to how charge costs are set to be able to set car specific settings? (add a battery type single stat chart to battery health that when clicking on it opens a settings page for the currently selected car in teslamate that then updates a value in the cars table - changing it should trigger recalculate_efficiency afterwards). 
> 
> @JakobLichterfeld - what do you think?

Are we turning circles? That's what is needed, right?

**Comment by JakobLichterfeld** [comment]
> Are we turning circles? 

Yes, at least the person asking the same question again and again.

**Comment by inigoml** [comment]
> Ok thank you. I'll charge like you suggest and let you know if something change in Teslamate. Regarding the best way to charge LFP battery, I read a lot, not sure that someone has the truth. So I ask Tesla directly who recommends to charge at 100% as often as possible, but yes, I should do it many times a week.

This has been commented widely. Tesla marketing area recommendw charging LFP battery to 100% in a regular basis because BMS miscalibration due to flat voltage curve in LFP chemistry, but this is not very good for degradation. First reports have arisen and show a higher than expected calendar aging to to this recommendation. As a former LFP owner, I can assure that this is not needed and recommendation should be "charge at 100% each 8 cycles" or similar. LFP also degrades due to high SOC, unfortunately. A much better approach in terms of calendar aging is "charge up to 70% in a regular basis and once a month up to 100% in order to recalibrate battery system".

![battery_age_degradation](https://github.com/teslamate-org/teslamate/assets/7486708/6f940c35-11c5-420f-8f70-913a6d9321f3)

**Comment by Viald** [comment]
Can you please share your sources, especially on the graphics?

**Comment by juneidy** [comment]
Hi everyone, I also am encountering the same issue (Model Y with LFP battery) most likely due to always charging to 100%.

Based on what I can understand, in order to get the dashboard display all values, we have to either

- wait for some form of https://github.com/teslamate-org/teslamate/issues/3524#issuecomment-1850452981 to be implemented, or
- only charge up to 95% for most of charging sessions and charge to 100% only once a week

Is that correct?

**Comment by SgtBill88** [comment]
Hello all, any advice on how to save or fix the useable battery level in the Battery Health page. No experience with Grafana. Thanks for the help.

**Comment by Viald** [comment]
> Is that correct?

Yes it is !

**Comment by Ali229** [comment]
![image](https://github.com/teslamate-org/teslamate/assets/11164440/49710139-ecc6-4749-91d7-91eaa474b9e8)
I have data since 5 Feb 2024, almost 2 months yet I don't see any battery data, I drive everyday and charge.

**Comment by jheredianet** [comment]
> .... I have data since 5 Feb 2024, almost 2 months yet I don't see any battery data, I drive every day and charge.

Hi, as I can see in the screenshot, some data is showing, so I guess that the "efficiency" is not being able to get. Could you try to see what the "Efficiency" dashboard shows?
There should be something like this:

![image](https://github.com/teslamate-org/teslamate/assets/45735191/67f0fd1b-c072-4386-83f8-a948c4e6d70f)

Teslamate, in order to get this information, needs some charges sessions that at least have more than 15kWh, more than 10 minutes and SOC less than 95%. Based on that data the Battery Health dashboards do the math and estimations. 

You should also check the "Charges" dashboard in order to browse if the charges are logged correctly, especially in the fields I've mentioned before:

![image](https://github.com/teslamate-org/teslamate/assets/45735191/f5882c9e-2868-4f4c-8186-c72f36dcfe15)

**Comment by Ali229** [comment]
Efficiency:
![image](https://github.com/teslamate-org/teslamate/assets/11164440/311463dd-21b4-4243-b1be-e51422446843)
Charges:
![image](https://github.com/teslamate-org/teslamate/assets/11164440/44a203c5-1717-4344-a595-3620436f919d)
Looks like all the data required is there no?

**Comment by jheredianet** [comment]
> Looks like all the data required is there, no?

Ok, this is what I meant in te previous message, as you see you only have short charges:

![image](https://github.com/teslamate-org/teslamate/assets/45735191/d815f070-0b4b-4974-99ef-6346daba14ca)

You have only added 8 kWh or 10 kWh as max on your charges. Based on your efficiency (219Wh/mi), you need at least a charge of more than 22kWh. You don't get low SOCs, try one day to charge from 20% to 80% for example.
